drv/transceivers-server/src/udp.rs

```diff
@@ -217,397 +215,572 @@ impl ServerImpl {
         &mut self,
         modules: ModuleId,
+        result: ModuleResult,
+        failure_type: Option<HwError>,
+        data_idx: usize,
         out: &mut [u8],
-    ) -> Result<usize, Error> {
-        let fpga = get_fpga(modules)?;
-        let fpga = self.transceivers.fpga(fpga);
-
-        // This is a bit awkward: the FPGA will get _every_ module's
-        // status (for the given FPGA), then we'll unpack to only the
-        // ones that we care about
-        let enable: U16<LittleEndian> =
-            fpga.read(Addr::QSFP_POWER_EN0).map_err(|_e| {
-                Error::StatusFailed(HwError::PowerEnableReadFailed)
-            })?;
-        let resetl: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_MOD_RESETL0)
-            .map_err(|_e| Error::StatusFailed(HwError::ResetLReadFailed))?;
-        let lpmode: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_MOD_LPMODE0)
-            .map_err(|_e| Error::StatusFailed(HwError::LpModeReadFailed))?;
-        let presentl: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_MOD_MODPRSL0)
-            .map_err(|_e| Error::StatusFailed(HwError::ModPrsLReadFailed))?;
-        let irql: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_MOD_INTL0)
-            .map_err(|_e| Error::StatusFailed(HwError::IntLReadFailed))?;
-        let pg: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_POWER_GOOD0)
-            .map_err(|_e| Error::StatusFailed(HwError::PgReadFailed))?;
-        let pg_to: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_POWER_GOOD_TIMEOUT0)
-            .map_err(|_e| Error::StatusFailed(HwError::PgTimeoutReadFailed))?;
-        let pg_lost: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_POWER_GOOD_LOST0)
-            .map_err(|_e| Error::StatusFailed(HwError::PgLostReadFailed))?;
-
-        // Write one bitfield per active port in the ModuleId
+    ) -> (usize, ModuleId) {
+        let mut error_idx: usize = data_idx;
+        // any modules at index 32->63 are not currently supported.
+        let invalid_modules = ModuleId(0xffffffff00000000);
+        let requested_invalid_modules = ModuleId(modules.0 & invalid_modules.0);
+        // iterate through the results and populate errors in the response buffer
+        // failure: whatever `HwError` specified by `failure_type`
+        // error: fpga communication issue
+        for module in modules.to_indices().map(LogicalPort) {
+            if module <= LogicalPortMask::MAX_PORT_INDEX {
+                if result.failure().is_set(module) {
+                    match failure_type {
+                        Some(e) => {
+                            out[error_idx] = e as u8;
+                            error_idx += 1;
+                        }
+                        None => {
+                            ringbuf_entry!(Trace::UnexpectedFailure(ModuleId(
+                                module.0 as u64
+                            )));
+                            panic!();
```

Comment:
I might add a useful panic message here. One could also make this `unreachable!()` which more strongly indicates the programmer definitely didn't intend this to be possible.

Reply:
I'll make this `unreachable!()` and elaborated in the comment above the function. My understanding is that in Hubris we try to be very sparing on including `panic!()` messages since that adds program text, so we try to include useful information in more compact ways.

Reply:
That makes sense in the constrained environment!